### ISSUE TITLE
The entity type IdentityRole is not part of the model for the current context.

### ISSUE THREAD

**Issue by ltraus** [IdentityServer/Thinktecture.IdentityServer3#838]
"The entity type IdentityRole is not part of the model for the current context."

I'm fairly sure I've done this, but i can't seem to figure out why, i've debugged and can't determine how I've messed this up.

Every time i try to log in, i get this in the stack.

    w3wp.exe Error: 0 :     [Thinktecture.IdentityServer.Core.Configuration.Hosting.ErrorPageFilterAttribute]: 1/28/2015 6:01:03   PM +00:00 -- Exception accessing: /login
    System.InvalidOperationException: The entity type IdentityRole is not part of the model for the    current context.
       at System.Data.Entity.Internal.InternalContext.GetEntitySetAndBaseTypeForType(Type entityType)
       at System.Data.Entity.Internal.Linq.InternalSet`1.Initialize()
       at System.Data.Entity.Internal.Linq.InternalSet`1.GetEnumerator()
       at System.Linq.Lookup`2.CreateForJoin(IEnumerable`1 source, Func`2 keySelector,    IEqualityComparer`1 comparer)
       at System.Linq.Enumerable.<JoinIterator>d__61`4.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
      at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
      at Microsoft.AspNet.Identity.EntityFramework.UserStore`6.GetRolesAsync(TUser user)
      at Microsoft.AspNet.Identity.UserManager`2.<GetRolesAsync>d__ac.MoveNext()
     --- End of stack trace from previous location where exception was thrown ---

I'm sure this is going to be a simple one but I can't seem to figure this out...something like this when   I've used aspnetidentity before was because i was not initializing the identity on the service, for a    custom call.

But going against your examples directly i'm at a loss.

I've configured the user service just as the examples.

    factory.Register(new Registration<Context>(resolver => new Context(connString)));
    factory.Register(new Registration<UserManager>());
    factory.Register(new Registration<UserStore>());
    factory.UserService = new Registration<IUserService, UserService>();

The identity manager is configured up for the RoleManager, but the IdentityManagerService has that capabitlity.

IS it that simple that the 2 context's are conflicting with each other? And i have to bridge the gap in the user service? 

Everything looks so standard i can't seem to figure out how i busted this up?

**Comment by ltraus** [comment]
I just had a brain fart and decided to go the other way; it's the entities classes; i switched over to the IntKey; for the user; and it all works fine can log in and such.

So the configuration for the UserManager and UserStore is busted up.

I'm probably going to go back and see if i can correct it now but it seems the sample doesn't work straight out?

**Comment by brockallen** [comment]
Sorry, I'm confused -- so is there a problem with the default/sample code, or was this just something misconfigured on your side?

**Comment by wilssoun** [comment]
I pondered this a while too. I figured out the problem might be that IdentityManager creates AspNetRoles table with string-typed Id while Asp.net Identity creates it with Int. So if you run the sample with simple entities, it won't work. Custom entities work fine.

**Comment by brockallen** [comment]
Ok, I'll fix that -- what did you do in the app to trigger the exception? IOW, what are the steps to repro?

**Comment by wilssoun** [comment]
I just used SimpleEntites and UserServices classes from the AspNetIdentity sample with WebHost sample. I managed to use the admintool properly but logging in with the server didn't work. When I switched to CustomEntities everything was ok. But as I said, the problem lies in roles table, I think the key shoudln't be GUID but Int. Thanks from awesome work btw :)

**Comment by brockallen** [comment]
ok, thanks for the update.

**Comment by ltraus** [comment]
Sorry for the delay but yeah its the identity, aspnetidentity without specification of the role on the user, it doesn't create the context.

SO when i added "string" just like wilssoun was mentioning it would work with int as well, it then created the ref context for the role.

But it enabled the ability to use the default GUID under the identitydbcontext.

     public class UserStore : UserStore<User, Role, string, IdentityUserLogin, IdentityUserRole,    IdentityUserClaim>
       {
           public UserStore(Context ctx)
              : base(ctx)
           {
           }
       }

Still not 100% sure if this is the most appropriate way about it, i was trying to dig up my old implementations of aspnetidentity out of the box, did something similar but didn't find it.

I'm actually working over all the examples building prototypes up for demonstration among the team i run, we're considering rolling it out between 2 of our software platforms.  So i'm "Demo" guy for the next few weeks ;)  So i'll let you know if i notice anything else.

**Comment by brockallen** [comment]
Ok, I updated the AspId sample to address this. Give it a try again (on dev), please.

**Comment by mrfoster** [comment]
I'm having the same issue with the latest version.  I managed to get it working briefly by removing the custom classes, fixing the compile errors and then using IdentityRole instead of Role.  I've now stripped it back even further to get rid of SimpleEntites and simply use the .net generic classes but have now ran into some IOC issue. I'll take another look if I get any time over the next few days.

**Comment by mrfoster** [comment]
Sorry I just noticed you changed the dev branch. Will check it out when I get home.

**Comment by subsci** [comment]
I verified that the new **SimpleEntities.cs** file (on dev) fixes the *IdentityRole not part of the model* #838.  Details next paragraph.  Wow great samples and tutorials and docs !

Note that the FirstName and LastName properties of the User (table, *AspNetUsers*) do not exist using the out-of-the-box MVC 5 VS2013 code first template. It seems the LastName and FirstName properties are referenced more than once in *UserService-AspNetIdentity*.  So I added these columns to the *AspNetUsers* table to get the example working completely. Note in this case the User Id is a GUID in a string column. As I used the MVC OWIN Client (Hybrid), I had to add the missing client definition for this client to the **Clients.cs** code of the *UserService-AspNetIdentity* (simple cut and paste from WebHost minimal sample).
